### ISSUE TITLE
[REVIEW]: The MCAPL Framework including the Agent Infrastructure Layer and Agent JavaPathfinder

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#617]
**Submitting author:** @louiseadennis (<a href="http://orcid.org/0000-0003-1426-1896">Louise Dennis</a>)
**Repository:** <a href="https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/" target ="_blank">https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/</a>
**Version:** 2017
**Editor:** @danielskatz
**Reviewer1:** @BasTesterink
**Reviewer2:** @fraimondi
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.1237635" target="_blank">10.5281/zenodo.1237635</a>

## Status

[![status](http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0/status.svg)](http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0"><img src="http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0/status.svg)](http://joss.theoj.org/papers/9f735950f5f68052fbda99e41f215cb0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer 1 instructions & questions

@BasTesterink, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @danielskatz know.

  ### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (2017)?
- [x] **Authorship:** Has the submitting author (@louiseadennis) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

## Reviewer 2 instructions & questions

@fraimondi, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @danielskatz know.

  ### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (2017)?
- [x] **Authorship:** Has the submitting author (@louiseadennis) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @BasTesterink it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #617 with the following error: 

 /app/vendor/ruby-2.3.4/lib/ruby/2.3.0/find.rb:43:in `block in find': No such file or directory (Errno::ENOENT)
	from /app/vendor/ruby-2.3.4/lib/ruby/2.3.0/find.rb:43:in `collect!'
	from /app/vendor/ruby-2.3.4/lib/ruby/2.3.0/find.rb:43:in `find'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-a31a5c2a9125/lib/whedon/processor.rb:52:in `find_paper_paths'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-a31a5c2a9125/bin/whedon:32:in `prepare'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-a31a5c2a9125/bin/whedon:99:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.3.0/bin/whedon:22:in `load'
	from /app/vendor/bundle/ruby/2.3.0/bin/whedon:22:in `<main>'

**Comment by danielskatz** [comment]
@arfon - can you compile this paper again, as you did in https://github.com/openjournals/joss-reviews/issues/573

Can you also make sure both reviewers (Reviewer1: @BasTesterink; Reviewer2: @fraimondi) are invited and able to access the checklist?

**Comment by danielskatz** [comment]
I'm also just noting here that @fraimondi may need until April 20 to perform this review

**Comment by arfon** [comment]
@louiseadennis - I'm afraid I'm unable to download the code from the repository. The `git clone` operation keeps timing out. Also, browsing the files, I don't seem to be able to find the `paper.md` and `paper.bib` files. Could you please link to them here?

**Comment by arfon** [comment]
> Can you also make sure both reviewers (Reviewer1: @BasTesterink; Reviewer2: @fraimondi) are invited and able to access the checklist?

Done.

**Comment by louiseadennis** [comment]
@arfon the paper.md and paper.bib files are in doc/joss_paper https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/doc/joss_paper/paper.bib and https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/doc/joss_paper/paper.md

I've no idea what is going on with sourceforge.  I can't clone either, though I can fetch and push.

**Comment by louiseadennis** [comment]
@arfon as far as I can work out, cloning is failing because of a memory problem at the sourceforge end.  I can't ssh into sourceforge at the moment to try the fixes suggested online (source forge says it has disable shell while it transitions to a new datacenter) and while there are some largish files in the repo (jars of libs etc.) I don't think I have loads of huge files that could be causing this.  I'm a little stumped how to proceed here but will look into it further on Thursday when I have more time.

**Comment by arfon** [comment]
Thanks @louiseadennis - I've managed to compile the paper using the direct links you provided here: [10.21105.joss.00617.pdf](https://github.com/openjournals/joss-reviews/files/1807583/10.21105.joss.00617.pdf)

**Comment by BasTesterink** [comment]
I have very little to comment on this submission. Within minutes I was able to download and compile the framework on my Windows 10 machine with an Eclipse Oxygen installation. The examples that I tried (there are a lot of them) all ran without any errors on the first try. The code is well documented, clean and readable, both for the core Java code and the example agent-programming code. The manual contains extensive explanations of the kind of software this is and discusses the examples on the different application domains (those being the different agent programming languages that the framework supports). 

I'm not sure what the expectations from JOSS are on community guidelines. There is no explicit guidelines sections in the readme for instance. I checked the review box because for me it seemed clear that any questions/suggestions should go to @loiseadennis or though the SF ticket system or mailing list. 

In conclusion, I believe the contribution to be a valuable piece of software for agent-oriented programmers. The MCAPL project is one of the few up-to-date initiatives on agent program verification which gets more and more relevant now that AI & Ethics is attracting attention.

**Comment by danielskatz** [comment]
Thanks @BasTesterink 

> I'm not sure what the expectations from JOSS are on community guidelines. There is no explicit guidelines sections in the readme for instance. I checked the review box because for me it seemed clear that any questions/suggestions should go to @loiseadennis or though the SF ticket system or mailing list.

There should be an explicit guidelines section somewhere, likely in the README. It's useful for people new to open source to have some text that tells them how they can ask questions, or how they can make contributions (both in terms of mechanism and policy).

**Comment by louiseadennis** [comment]
Not 100% sure of the process here.  But I've added a "Problems and Contributions" section to the README and pushed to the master branch.

**Comment by danielskatz** [comment]
For some reason, https://sourceforge.net/p/mcapl/mcapl_code/ci/master/tree/ isn't currently working

**Comment by danielskatz** [comment]
now it's back - not sure what happened...

**Comment by louiseadennis** [comment]
It was behaving strangely when @arfon was looking at it on Wednesday, throwing errors which implied a memory management problem at some level.  I was going to look into it but then the problem went away so I assumed sourceforge had fixed it at their end.

**Comment by danielskatz** [comment]
Just adding a note that @fraimondi's review is expected the 20th of April - we'll be on hold until then.

**Comment by fraimondi** [comment]
I have done my review and I am happy with the repository, the documentation and the paper. The installation and examples all worked first time. In addition to Eclipse, I tried from the command line in Linux and Mac: no problem .

Very minor point (may be useful for bots?): the LICENSE file is called COPYING, is this OK?

**Comment by danielskatz** [comment]
The license file should be called LICENSE for consistency with general practices and user/developer expectations.  @louiseadennis - can you fix this?  Then we can move forward with acceptance.

**Comment by danielskatz** [comment]
and thanks @fraimondi !

**Comment by danielskatz** [comment]
👋 @louiseadennis - note this is waiting for a relatively small change from you.

**Comment by louiseadennis** [comment]
It's on the stack, we're just in the middle of project marking here.  Should get to it this week or early next week at the very latest.

**Comment by louiseadennis** [comment]
Changed and pushed to the repository!

**Comment by danielskatz** [comment]
@fraimondi - I thought this was ready to be accepted, but my final look shows that there is one box left unchecked - yours on functionality.

Can you tell us what you think is needed?

**Comment by fraimondi** [comment]
Apologies, my mystake! The only thing that is needed is a new pair of glasses for the reviewer.
I've now checked all the boxes, I think it is ready to be accepted.
Thanks!

**Comment by danielskatz** [comment]
Great!
👋 arfon - can we move this to acceptance now?

**Comment by danielskatz** [comment]
👋 @arfon - can we move this to acceptance now?

**Comment by arfon** [comment]
> Is that sufficient or do I need to generate a new archive altogether?

This should be good.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.1237635 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.1237635" target="_blank">10.5281/zenodo.1237635</a> is the archive.

**Comment by arfon** [comment]
@BasTesterink, @fraimondi - many thanks for your reviews here and to @danielskatz for editing this submission ✨ 

@louiseadennis - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00617 ⚡️:rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippet:

```
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00617/status.svg)](https://doi.org/10.21105/joss.00617)
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00617/status.svg)](https://doi.org/10.21105/joss.00617)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by louiseadennis** [comment]
I just noticed a typo in the title of the paper (an instead of and).  I've fixed in the repository.  Is there anyway this can be fixed online?

**Comment by arfon** [comment]
> I just noticed a typo in the title of the paper (an instead of and). I've fixed in the repository. Is there anyway this can be fixed online?

I've just recompiled the paper so this should be fixed now.

**Comment by louiseadennis** [comment]
Thanks!
